zerocopy-derive/src/lib.rs

```diff
@@ -634,6 +635,47 @@ fn derive_from_zeros_struct(ast: &DeriveInput, strct: &DataStruct) -> proc_macro
 // - all of its variants are fieldless
 // - one of the variants has a discriminant of `0`
 
+fn find_zero_variant<'a>(variants: impl IntoIterator<Item = &'a Variant>) -> Option<usize> {
+    let mut current_implicit_discriminant = Some(0isize);
+    for (i, v) in variants.into_iter().enumerate() {
+        // Whenever we encounter a discriminant that we don't understand, we
+        // stop tracking the current implicit discriminant. Those could have any
+        // value, even zero or a negative number.
+        match v.discriminant.as_ref() {
+            // Implicit discriminant
+            None => {
+                if current_implicit_discriminant == Some(0) {
+                    return Some(i);
+                }
+            }
+            // Explicit discriminant
+            Some((_, Expr::Lit(ExprLit { lit: Lit::Int(int), .. }))) => {
+                match int.base10_parse::<usize>().ok() {
+                    Some(0) => return Some(i),
+                    Some(x) => current_implicit_discriminant = Some(x as isize),
+                    None => current_implicit_discriminant = None,
+                }
+            }
+            // Explicit negative discriminant
+            Some((_, Expr::Unary(ExprUnary { op: UnOp::Neg(_), expr, .. }))) => match &**expr {
+                Expr::Lit(ExprLit { lit: Lit::Int(int), .. }) => {
+                    match int.base10_parse::<usize>().ok() {
+                        Some(0) => return Some(i),
+                        Some(x) => current_implicit_discriminant = Some(-(x as isize)),
+                        None => current_implicit_discriminant = None,
+                    }
+                }
+                _ => current_implicit_discriminant = None,
+            },
+            _ => current_implicit_discriminant = None,
+        }
+        if let Some(n) = current_implicit_discriminant.as_mut() {
+            *n += 1;
```

Comment:
Can we bounds check this to make sure it won't overflow?

Reply:
Done, same as above